### PR TITLE
[CLOUDSTACK-10039] Adding IOPS/GB offering

### PR DESCRIPTION
**Pull request by syed**
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

We want to add a disk offering where we can specify the Min/Max IOPS as a function of size. The idea is the larger volume you create, the greater your IOPS will be (this is similar to the way GCE handles IOPS). We also have added limits to the IOPS (min and max) so that once the volume goes beyond a certain size, the IOPS won't change and are capped to the given values.

The following parameters are added to the `createDiskOffering` API:

1. `miniopspergb' : Minimum IOPS/GB for the offering
2. `maxiopspergb`: Maximum IOPS/GB for the offering
3. `highestminiops`: The highest `miniops` value that is allowed for this offering
3. `highestmaxiops`: The highest `maxiops` value that is allowed for this offering

### CODE REVIEW

**Comment by wido** [comment]
Could you resolve the conflict @syed ?

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 1126
 Hypervisor xenserver
 NetworkType Advanced
 Passed=109
 Failed=5
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_router_dnsservice.py

 * test_router_dns_guestipquery Failed

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 124 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 124 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 120 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 120 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by rhtyd** [comment]
@syed can you fix the conflicts?

**Comment by rhtyd** [comment]
@syed ping

**Comment by syed** [comment]
@rhtyd Sorry Rohit, looks like this slipped out of my radar. I've rebased and resolved the conflicts.

**Comment by rhtyd** [comment]
@syed looks like we have new conflicts

**Comment by syed** [comment]
darn it! I'll resolve it now

**Comment by syed** [comment]
@rhtyd Done!

**Comment by bwsw** [comment]
Cool feature. Recently have thought about it.

**Comment by rhtyd** [comment]
@syed can you fix the conflicts?

**Comment by rafaelweingartner** [comment]
Ping @syed

**Comment by syed** [comment]
@rafaelweingartner I've resolved the conflicts. I'd really like to get this in because I think this is the 3rd time rebasing it. If you could do a code review, that would be sublime :)

File: engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql [hunk 1]
```diff
@@ -0,0 +1,70 @@
+-- IOPS per GB disk offering
+DROP PROCEDURE IF EXISTS `cloud`.`ADD_COLUMN_TO_TABLE_IDEMPOTENT`;
+
+DELIMITER //
+CREATE PROCEDURE `cloud`.`ADD_COLUMN_TO_TABLE_IDEMPOTENT`(
```
Comment (rafaelweingartner) [inline engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql]: Isn't this being introduced by some other PR?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -2566,6 +2567,61 @@ protected DiskOfferingVO createDiskOffering(final Long userId, final Long domain
             isCustomized = true;
         }
 
+        if (minIopsPerGb != null || maxIopsPerGb != null) {
+
+            if (!isCustomized) {
+               throw new InvalidParameterValueException("Cannot set Min/Max IOPS/GB for a fixed size disk offering");
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: I did not understand this error here. Can you explain it a little further?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -2566,6 +2567,61 @@ protected DiskOfferingVO createDiskOffering(final Long userId, final Long domain
...
+            }
+
+            if ((isCustomizedIops != null && isCustomizedIops) || minIops != null || maxIops != null) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: Can you use BooleanUtils here?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -2566,6 +2567,61 @@ protected DiskOfferingVO createDiskOffering(final Long userId, final Long domain
...
+                throw new InvalidParameterValueException("Cannot set Min/Max IOPS/GB with either " +
+                        "custom IOPS or fixed IOPS");
+            }
+
+            if (minIopsPerGb != null && maxIopsPerGb != null) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: Could you extract this block to a specific method? Document it, and write unit tests for it?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 5]
```diff
@@ -2566,6 +2567,61 @@ protected DiskOfferingVO createDiskOffering(final Long userId, final Long domain
...
+                }
+            }
+
+            //if either one of them is set but the other is not
+            if ((minIopsPerGb != null && maxIopsPerGb == null) || (minIopsPerGb == null && maxIopsPerGb != null)) {
+                throw new InvalidParameterValueException("Both min IOPS/GB and max IOPS/GB must be specified");
+            }
+        }
+
+        if (highestMinIops != null && highestMaxIops != null) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: This whole validation block can also be extracted to a method, documented and unit tested.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 6]
```diff
@@ -916,29 +951,34 @@ public VolumeVO resizeVolume(ResizeVolumeCmd cmd) throws ResourceAllocationExcep
                 // no parameter provided; just use the original size of the volume
                 newSize = volume.getSize();
             }
-
+            newSizeInGb = newSize >> 30;
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Let's not use bit shift here. It makes things harder to read and understand.

The code is already convoluted enough

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 7]
```diff
@@ -981,6 +1021,7 @@ public VolumeVO resizeVolume(ResizeVolumeCmd cmd) throws ResourceAllocationExcep
                 newSize = newDiskOffering.getDiskSize();
             }
 
+            newSizeInGb = newSize >> 30;
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: bit shift...

Have you checked if there is a library to convert between these bases? If there is none, what about we create an utils method to execute this job?

File: setup/db/db/schema-41000to41100.sql [hunk 8]
```diff
@@ -0,0 +1,324 @@
+-- Licensed to the Apache Software Foundation (ASF) under one
+-- or more contributor license agreements.  See the NOTICE file
```
Comment (rafaelweingartner) [inline setup/db/db/schema-41000to41100.sql]: these scripts should not be here, right?

File: engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql (hunk 1 above)
Comment (syed) [reply engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql]: No. This is only for this PR

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 2 above)
Comment (syed) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: You cannot have an IOPS/GB offering for a fixed sized disk. The whole Idea of an IOPS/GB type of offering makes sense when you have a custom offering so when you change the size of a disk, the IOPS also change

File: engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql (hunk 1 above)
Comment (rafaelweingartner) [reply engine/schema/src/main/resources/META-INF/db/schema-41100to41200.sql]: Hmm.. Then, I think it is better to coordinate with PR #2449.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 2 above)
Comment (rafaelweingartner) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: So, for fixed size disks we cannot control the IOPS?  Why cannot we control/limit the IOPS of fixed size disks?

BTW, what do you understand as IOPS? input and output per second, meaning read and write speed.

**Comment by bwsw** [comment]
@syed Hi, do you continue to manage this PR? I think it's very beneficial for 4.12.

**Comment by svenvogel** [comment]
@syed is it useful for solidfire? Is there any progress in this pr?

**Comment by rafaelweingartner** [comment]
@syed I will move this to 5.0.0. The freeze date is coming, and it does not seem that we will be able to add it to 4.12.0.

**Comment by GabrielBrascher** [comment]
@rhtyd @syed @DaanHoogland moving this to 4.15. Still pending review and fixing conflicting files.

**Comment by GutoVeronezi** [comment]
@syed this is an interesting feature; Do you intend to continue managing this PR?

**Comment by GabrielBrascher** [comment]
I think that this PR is no longer necessary due to PR adding support for IOPS burst https://github.com/apache/cloudstack/pull/3133.

**Comment by nvazquez** [comment]
Hi @syed can you please confirm if this PR is still needed or can be closed by PR3133? In case it is, is it ready for review?

**Comment by syed** [comment]
Hi @nvazquez Sorry for the delayed response. #3313 is different from this PR. The idea here is to give the operator the ability to expose storage offerings in the style AWS gp2/gp3  where IOPS scales with size.  I currently don't have a setup to work on this but would be happy to support if someone picks it up

**Comment by nvazquez** [comment]
Thanks @syed this will need some effort and resolving many conflicts. Do you think you can pick this up sometime soon? Otherwise we can move it for the next milestone

**Comment by syed** [comment]
Probably good idea to move to next milestone

**Comment by rohityadavcloud** [comment]
Hi @syed thank you for your PR, given the inactivity in the last 4+ years, I'm closing this. If you happen to work on this feature and can fix conflicts pl re-open or re-raise the PR.

**Pull request closed without merging**